### PR TITLE
NXP-31126: introduce compound documents breadcrumb enricher

### PR DESCRIPTION
**Pull request by NourNuxeo**
New Compound Document Breadcrumbs to allow searching compounds by the tags and full text of their assets as aggregates.

### CODE REVIEW

**Comment by NourNuxeo** [comment]
flattening the loops would make it clearer but more verbose. 
The test is modeled on https://github.com/nuxeo/nuxeo-lts/blob/2021/modules/core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/io/marshallers/json/enrichers/BreadcrumbJsonEnricherTest.java

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            }
+            parentDocuments = wrapper.getSession().getParentDocuments(document.getRef());
+        }
+        var parentCompounds = parentDocuments.stream()
+                                             .filter(d -> d.hasFacet(COMPOUND_DOCUMENT_FACET))
+                                             .collect(Collectors.toList());
+        DocumentModelListImpl documentList = new DocumentModelListImpl(parentCompounds);
+        jg.writeFieldName(NAME);
+        writeEntity(documentList, jg);
+    }
```
Comment (troger) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java]: Why not simply something like:

```suggestion
    public void write(JsonGenerator jg, DocumentModel document) throws IOException {
        try (SessionWrapper wrapper = ctx.getSession(document)) {
            if (!wrapper.getSession().exists(document.getRef())) {
                return;
            }
            var docs = wrapper.getSession().getParentDocuments(document.getRef())
                                           .stream()
                                           .filter(d -> d.hasFacet(COMPOUND_DOCUMENT_FACET))
                                           .collect(Collectors.toList());
            jg.writeFieldName(NAME);
            writeEntity(docs, jg);
        }
    }
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Before
+    public void setup() {
+        document = session.getRootDocument();
+        for (int i = 1; i <= LEVELS; i++) {
+            String type = i % 2 == 0 ? COMPOUND_FOLDER_DOCTYPE : COMPOUND_DOCTYPE;
+            document = session.createDocumentModel(document.getPathAsString(), "level" + i, type);
+            document = session.createDocument(document);
+        }
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: I would separate the created hierarchy from an asset document inside the hierarchy.
`document` should only be used to create the "asset" that we will be testing in the other method.

Quickly reading the code, I don't know what `document` will contain between a folder or a compound - and we probable want to test it on an asset such as a File/Note/whatever.

> flattening the loops would make it clearer but more verbose.

I would say yes, we prefer clarity over less verbosity... especially here where we can just have a simple hierarchy such as:
```
compound doc > coumpound folder > coumpound doc > File
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        json.properties(1);
+        json = json.has("compoundDocumentBreadcrumb").isObject();
+        json.has("entity-type").isEquals("documents");
+        int expectedLevels = (int) Math.ceil((double) LEVELS / 2);
+        json = json.has("entries").length(expectedLevels);
+        for (int i = 0, j = 0; i < 3; i += 2, j++) {
+            JsonAssert doc = json.has(j);
+            doc.has("entity-type").isEquals("document");
+            doc.has("title").isEquals("level" + (i + 1));
+        }
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: Reading that convinced me for more clarity :)

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java (hunk 1 above)
Comment (NourNuxeo) [reply nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java]: I wanted to do the filtering outside the try resources to keep it to a minimum.
Ok I'll do that.

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        document = session.createDocument(document);
+        document = session.createDocumentModel("/level1/level2", "level3", COMPOUND_DOCTYPE);
+        document = session.createDocument(document);
+        document = session.createDocumentModel("/level1/level2/level3", "level4", "File");
+        document = session.createDocument(document);
+    }
+
+    @Test
+    public void test() throws IOException {
+        JsonAssert json = jsonAssert(document, CtxBuilder.enrichDoc("compoundDocumentBreadcrumb").get());
```
Comment (troger) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: Use the constant here maybe.

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                return;
+            }
+            var parentCompounds = wrapper.getSession()
+                                         .getParentDocuments(document.getRef())
+                                         .stream()
+                                         .filter(d -> d.hasFacet(COMPOUND_DOCUMENT_FACET))
+                                         .collect(Collectors.toList());
+            DocumentModelListImpl documentList = new DocumentModelListImpl(parentCompounds);
+            jg.writeFieldName(NAME);
+            writeEntity(documentList, jg);
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentBreadcrumbJsonEnricher.java]:
```suggestion
            var parentCompounds = wrapper.getSession()
                                         .getParentDocuments(document.getRef())
                                         .stream()
                                         .filter(d -> d.hasFacet(COMPOUND_DOCUMENT_FACET))
                                         .collect(Collectors.toCollection(DocumentModelListImpl::new));
            jg.writeFieldName(NAME);
            writeEntity(parentCompounds, jg);
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    public void setup() {
+        document = session.createDocumentModel("/", "level1", COMPOUND_DOCTYPE);
+        document = session.createDocument(document);
+        document = session.createDocumentModel("/level1", "level2", COMPOUND_FOLDER_DOCTYPE);
+        document = session.createDocument(document);
+        document = session.createDocumentModel("/level1/level2", "level3", COMPOUND_DOCTYPE);
+        document = session.createDocument(document);
+        document = session.createDocumentModel("/level1/level2/level3", "level4", "File");
+        document = session.createDocument(document);
+    }
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: I would be in favor to not do a setup step as there's only one test, and put the document creation directly in test with some naming change like `level1Doc`, `level2Doc`, etc...

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java (hunk 4 above)
Comment (NourNuxeo) [reply nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: I also used it in the has a few lines below

**Review by kevinleturc — COMMENTED** [review]
Review not taken into account

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    public void test() throws IOException {
+        DocumentModel document = session.createDocumentModel("/", "level1Doc", COMPOUND_DOCTYPE);
+        session.createDocument(document);
+        document = session.createDocumentModel("/level1Doc/", "level2Doc", COMPOUND_FOLDER_DOCTYPE);
+        session.createDocument(document);
+        document = session.createDocumentModel("/level1Doc/level2Doc", "level3Doc", COMPOUND_DOCTYPE);
+        session.createDocument(document);
+        document = session.createDocumentModel("/level1Doc/level2Doc/level3Doc", "level4Doc", "File");
+        document = session.createDocument(document);
+        JsonAssert json = jsonAssert(document, CtxBuilder.enrichDoc(NAME).get());
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: My suggestion about the renaming was on the variable not the document title, as it we know exactly with no doubt which document is asserted, so we will end with something like (added a new line to ease reading):
```suggestion
        var level1doc = session.createDocumentModel("/", "level1", COMPOUND_DOCTYPE);
        session.createDocument(level1doc);
        var level2Doc = session.createDocumentModel("/level1/", "level2", COMPOUND_FOLDER_DOCTYPE);
        session.createDocument(level2Doc);
        var level3Doc = session.createDocumentModel("/level1/level2", "level3", COMPOUND_DOCTYPE);
        session.createDocument(level3Doc);
        var level4Doc = session.createDocumentModel("/level1/level2/level3", "level4", "File");
        level4Doc = session.createDocument(level4Doc);
        
        JsonAssert json = jsonAssert(level4Doc, CtxBuilder.enrichDoc(NAME).get());
```

Comment (NourNuxeo) [reply nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentBreadcrumbJsonEnricher.java]: Arf I'm sorry

**Review by kevinleturc — APPROVED** [review]

**Review by troger — APPROVED** [review]

**Pull request merged**